### DISCUSSION TITLE
Fixture convex vertices not changing when object moves

### DISCUSSION THREAD

**Discussion by ProgrammingMuffin** [dyn4j/dyn4j#240] [Q&A] [answered]
Hi all,

I made these changes in `bucket.java` and in `SimulationFrame.java` in dyn4j-samples to see if convex vertices change.
Ignore "Something" class. It's a superglobal I am using for testing this out and to block out other bodies in the world.
 
*Bucket.java*
```java
this.bucketMovable = new SimulationBody();
bucketMovable.addFixture(Geometry.createRectangle(15.0, 1.0));
bucketMovable.setMass(MassType.NORMAL);
bucketMovable.setUserData(new Something());
Something.bucketMovable = this.bucketMovable;
world.addBody(bucketMovable);
```

*SimulationFrame.java*
```java
public static final void render(Graphics2D g, Polygon polygon, double scale, Color color) {
	Vector2[] vertices = polygon.getVertices();
	int l = vertices.length;

	if (Something.bucketMovable.getFixture(0).getShape().equals(polygon)) {
		System.out.print("fixtures: ");
		for (Vector2 vec : vertices) {
			System.out.println(vec);
		}
	}
```
This is called inside renderFixtures() method.

My output is like this:

```
(7.5, -0.5)
(7.5, 0.5)
(-7.5, 0.5)
fixtures: (-7.5, -0.5)
(7.5, -0.5)
(7.5, 0.5)
(-7.5, 0.5)
fixtures: (-7.5, -0.5)
(7.5, -0.5)
(7.5, 0.5)
(-7.5, 0.5)
```

What doesn't match is that the fixture actually moved but it doesn't reflect in the print statement here. Any idea what's going on here? Am I missing something?

All help appreciated.

**Comment by wnbittle** [comment, marked as answer]
There are a few layers in dyn4j: `Body`, `BodyFixture`, and `Convex`.  The `Body` is what is being simulated.  The `BodyFixture` is where physical properties exist for each `Convex`.  The `Convex` just represents it's shape.  The `Body` is used to group multiple `Convex` shapes to create a non-convex shape (if desired) and because it's the object being simulated, not the individual convex shapes, the `Body` stores the true position and rotation.  It stores this in a coordinate `Transform`.

All that said, imagine that each `Body` defines it's own local coordinate system.  The `Transform` can be used to convert from _local body_ coordinates to _world_ coordinates and the other way around.  So if you want to know the true location of the vertices of a `Convex` attached a `Body` you need to transform them to world coordinates:

```java
Body b = ...;
Transform tx = b.getTransform();
// NOTE: not every shape extends the Wound interface - this is just an example
Wound w = (Wound)b.getFixture(0).getShape();
Vector2 localPoint = w.getVertices()[0];
Vector2 worldPoint = tx.getTransformed(localPoint);
// or 
// Vector2 worldPoint = b.getWorldPoint(localPoint);
```

**Reply by ProgrammingMuffin** [reply]
Thank you, this has given me an understanding of how everything fits.
